Create dependencies for NETStandard.Library facades

tmeschter opened the pull request:
The NETStandard.Library package (starting from version 2.0) adds
references to most assemblies via a .targets file rather than going
through the normal process to resolve assemblies from packages. This
causes the assemblies to show up under the Assemblies node in Solution
Explorer, rather than under the NETStandard.Library package where they
logically belong.

To surface these assemblies under the package we need to update the
existing `FileDefinition` items for the assemblies, and create new
`FileDependency` items linking them to the package.

Note that this change does nothing to hide the items under the
Assemblies node; that will need to occur as a separate change in the
project system. This is also a targeted at the NETStandard.Library, and is
not a general mechanism that other packages could use to do the same
thing. If that is desired, we will need to design and implement it
separately.

**Customer scenario**

Customer creates a project targeting .NET Standard 2.0 and then starts looking through the Dependencies node in Solution Explorer. The references brought in by the NETStandard.Library package appear under the "Assemblies" node rather than under the "SDK\NETStandard.Library" node as expected.

**Bugs this fixes:** 

Partial fix for https://github.com/dotnet/project-system/issues/2320.

**Workarounds, if any**

None

**Risk**

Low.

**Performance impact**

This will reduce UI delays and reduce the number of active threads.

**Is this a regression from a previous update?**

Yes, in the sense that it doesn't occur with lower versions of .NET Standard.

**Root cause analysis:**

We only want to bring in the references from NETStandard.Library 2.0.0 if the project is actually targeting .NET Standard v2.0; if we're targeting something compatible with .NET Standard v2.0 then that target is responsible for providing the proper assembly references. Hence the NETStandard.Library package uses a custom .targets file to supply the `Reference` items rather than letting them be resolved from the package contents as you would normally expect. This causes them to show up in the wrong place in Solution Explorer.

**How was the bug found?**

Ad hoc testing.

tmeschter commented:
@dotnet/project-system Please review.

livarcocc commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -163,6 +176,55 @@ protected override void ExecuteCore()
+        {
+            var facades = References.Where(item => item.GetBooleanMetadata("Facade") == true
+                                                   && item.GetMetadata("NuGetPackageId") == "NETStandard.Library"
+                                                   && item.GetMetadata("NuGetPackageVersion") != null);
```
Should this be version >= 2.0?

tmeschter replied:
Possibly. But I wouldn't expect assemblies from older versions to have the Facade metadata.

livarcocc commented:
Can we add some unit tests for this?

srivatsn commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -163,6 +176,55 @@ protected override void ExecuteCore()
+        /// </remarks>
+        private void FixUpNetStandardLibraryFacadeAssemblies()
+        {
+            var facades = References.Where(item => item.GetBooleanMetadata("Facade") == true
```
It's not just facades - netstandard.dll for example won't have the Façade metadata but that needs to be reparented as well. Also why do we need to explicit check for NETStandard.Library? Can we just reparent items with NuGetPackageId and NuGetPackageVersion metadata under that package id,version?

tmeschter replied:
Why wouldn't netstandard.dll have the façade metadata?

tmeschter replied:
We could do this for other assemblies, except then I'm not sure what target they should belong to.

nguerrera replied:
> Why wouldn't netstandard.dll have the facade metadata?

Because this is the thing that provides all the API to a netstandard 2.0 library, not just a bunch of type forwarders for compat with other targets. Facade indicates the latter.

It should not be tied to NetStandard.Library. The package is using targets because it has different rules than nuget defaults about which TFMs actually consume its assets, so it uses targets to add references instead of the default nuget rules based on folders. Another package may come along and have the same requirement

We should reparent any Reference with NuGetPackageId under the corresponding package. Facade==true should be strictly about de-emphasizing or hiding facades should we decide to do so.

tmeschter commented:
@livarcocc I'm trying to write a unit test, but working in the dotnet/SDK repo is painful as there is no documentation.

natidea commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -163,6 +176,55 @@ protected override void ExecuteCore()
+                    .Where(def => def.GetMetadata(MetadataKeys.PackageName) == packageName
+                                  && def.GetMetadata(MetadataKeys.PackageVersion) == packageVersion
+                                  && def.GetMetadata(MetadataKeys.ResolvedPath) == facadePath)
+                    .FirstOrDefault();
```
>.FirstOrDefault(); [](start = 20, length = 18)

The first two conditions of the "Where", as well as packageName and packageVersion, can be lifted out of the for-loop, so you are not recomputing this information for each facade (since presumable they are all part of one package). 
Also the string comparisons should be Ordinal-Ignore-Case

tmeschter replied:
@natidea  It sounds like we don't want this to be tied to NETStandard.Library, so the package and version might be different.

I'll fix the string comparisons.

natidea commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -170,6 +170,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     <ResolvePackageDependencies
       ProjectPath="$(MSBuildProjectFullPath)"
       ProjectAssetsFile="$(ProjectAssetsFile)"
+      References="@(Reference)"
```
>References="@(Reference)" [](start = 6, length = 25)

Are we using this approach because the list of references is always strictly smaller than the list of DLLs listed in the assets file under NETStandard.Library? (Otherwise we could just mark all those assemblies as dependencies and avoid this parameter). Also, what ordering requirement guarantees that ResolvePackageDependencies is always called after the target the sets the Netstandard library references?

tmeschter replied:
@natidea We need the metadata from the Reference (that is, whether or not it is a façade, and the NuGetPackageId and NuGetPackageVersion).

I'll take a look at the ordering.

natidea commented:
>       <Reference Include="@(ResolvedCompileFileDefinitions)" />

We add these FileDependencies (with FileGroup='CompileTimeAssembly') back to the list of References. So presumably these facades would end up being included twice. I assume that is OK?

---
Refers to: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.PackageDependencyResolution.targets:420 in fb9b4c8. [](commit_id = fb9b4c8c6eaa3030abd751c74796d9d88fc82dd7, deletion_comment = False)

natidea commented:
You can unit test this by using the LockFileSnippet methods (CreateLibrary and CreateTargetLibrary) to construct a sample assets file with entries for NETStandard.Library, then checking the resulting file definitions and file dependencies for the right content

---
In reply to: [307514652](https://github.com/dotnet/sdk/pull/1330#issuecomment-307514652) [](ancestors = 307514652)

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -163,6 +176,55 @@ protected override void ExecuteCore()
+                var packageVersion = facade.GetMetadata("NuGetPackageVersion");
+                var facadePath = Path.GetFullPath(facade.ItemSpec);
+
+                var fileDefinition = _fileDefinitions
```
There are many facades and many file definitions, right. This N*M loop seems like it could be quite expensive. ResolvePackageDependencies is already showing up as a bottleneck.

tmeschter replied:
Yes, that will need to be addressed. First order of business is to get something that works correctly. :-)

nguerrera reviewed (changes requested):
See above

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -163,6 +176,55 @@ protected override void ExecuteCore()
+        /// create the appropriate file definitions and dependencies to associate them
+        /// with the package.
+        /// Here we find the "facade" references from the .NETStandard.Library package,
+        /// update the existing FileDefinition, and create a FileDependency between the
```
Sri, I thought you said the items used for the tree were distinct from these. I don't think rendering the IDE tree differently should impact items that are inspected by build.

tmeschter replied:
Different parts of the tree use different items. The nodes shown under Packages and SDK are based in part on the `FileDefinition` and `FileDependency` items.

nguerrera commented:
> @livarcocc I'm trying to write a unit test, but working in the dotnet/SDK repo is painful as there is no documentation.

You can run tests in 3 simple ways:

1. build.cmd (also runs all tests)
2. Set unit test as startup project, F5
3. Test explorer (if you can get past the perf issues with it)

All of these should work without any custom setup. Preview1+ VS should just work. All the complicated stuff we talked about on Teams was for getting VS to pick up a freshly built SDK. It does not apply to unit tests.

tmeschter commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageDependencies.cs:
```diff
@@ -163,6 +176,55 @@ protected override void ExecuteCore()
+            {
+                var packageName = facade.GetMetadata("NuGetPackageId");
+                var packageVersion = facade.GetMetadata("NuGetPackageVersion");
+                var facadePath = Path.GetFullPath(facade.ItemSpec);
```
@natidea The goal is to match up a FileDefinition with the corresponding Reference. I'd like to use the full path to the file to do that. The call to `Path.GetFullPath` is meant to "clean up" the path and remove extraneous `\`, `..`, etc. Would ResolveFilePath be a better choice for that?

tmeschter commented:
@natidea @nguerrera I got unit tests running eventually, but it took @dsplaisted's help and the better part of a day. :-(

nguerrera commented:
What was the issue with test running? Maybe trying to run with env vars set to pick up built sdk? I should have mentioned that these are exclusive. Tests should just work with no environment customization.

tmeschter commented:
@natidea 

>We add these FileDependencies (with FileGroup='CompileTimeAssembly') back to the list of References. So presumably these facades would end up being included twice. I assume that is OK?

You mean in addition to the existing `Reference` items, we take a subset of the `FileDependency` items and create new `Reference` items? That probably explains why this change causes all the items under the "Assemblies" node to show up with warning icons--we're getting duplicates.

Where do the `FileDependency` items get turned into `Reference` items? Your comment didn't attach itself properly to any line of code.

tmeschter committed:
Create dependencies for NETStandard.Library facades

The NETStandard.Library package (starting from version 2.0) adds
references to most assemblies via a .targets file rather than going
through the normal process to resolve assemblies from packages. This
causes the assemblies to show up under the Assemblies node in Solution
Explorer, rather than under the NETStandard.Library package where they
logically belong.

To surface these assemblies under the package we need to create new
`FileDependency` items linking them to the package. This is done in
the `PreprocessPackageDependenciesDesignTime` task as we don't want
to modify the normal build in any way, just surface some additional
information in the design-time build.

Note that this change does nothing to hide the items under the
Assemblies node; that will need to occur as a separate change in the
project system.

tmeschter commented:
@srivatsn @natidea @livarcocc @nguerrera Could you take another look at this? I've updated the PR with a different approach that only affects the design-time build.

tmeschter committed:
Fix up unit tests; add a new one

1. Fix up the existing unit tests to understand that the `PreprocessPackageDependenciesDesignTime` task now takes in a set of `References` items.
2. Add a basic unit test covering the workaround for assemblies in NETStandard.Library v2.0.0.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -182,6 +191,82 @@ private void PopulateAssemblies()
+        ///
+        /// This is not meant to be a general mechanism for injecting files that can't be handled
+        /// through the normal NuGet package mechanisms. For that, package authors should be
+        /// creating FileDefinition and FileDependency items instead of Reference items.
```
I still don't want to encourage manipulating FileDefinition and FileDependency here as I'm still convinced that we want to remove FileDefinition and FileDependency from all build logic and only promise that the output items of this task remain. There is a lot of perf to be gained if we don't produce all these items on every build.

I'd rather bless this path as the general mechanism. Basically, if you're not adding references from the assets file, then don't fake up the assets file data structure and pretend like you are. Instead, just annotate who added the `Reference`s the old-school way, and we'll take of putting together a tree view with both kinds of references.

srivatsn replied:
I agree - I see nothing wrong with supporting a NuGetPackageId\Version metadata property.

tmeschter replied:
@nguerrera Sorry, the comment pre-dates our conversation on Teams yesterday. I could simply remove the references to `FileDefinition` and `FileDependency`.

nguerrera replied:
Please do clean up the comment to not suggest FileDefinition and FileDependency approach.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -182,6 +191,82 @@ private void PopulateAssemblies()
+                if (packageName != "NETStandard.Library")
+                {
+                    continue;
+                }
```
Is this the only thing preventing it from generalizing? You mentioned something offline about having to set the TFM. Is that elsewhere or is it not a problem after all?

tmeschter replied:
@nguerrera @srivatsn Conceptually, we need three pieces of metadata to properly determine where a `Reference` should end up: the package name, package version, and target framework. The last is necessary for multi-targeting scenarios. The NETStandard.Library package provides `NuGetPackageId` and `NuGetPackageVersion` but not the target framework. So no matter what we do we will need to have some NETStandard.Library-specific workaround to inject that info.

I see a couple of issues with turning this into a general mechanism:

1. We need to extend it to other item types. Definitely `AnalyzerReference`, but what about other item types?
2. Specifying the target framework correctly might be tricky for package authors.

tmeschter replied:
Let me expand a bit on that second issue.

Say you are a package author and can't add references from the assets file for whatever reason so you need to use this mechanism. Say you also need to include different files for netstandard2.0 and net46. You'll need to do something like this:

``` xml
<Reference Include="net46support\Alpha.dll" Condition="'$(TargetFramework)' == 'net46'">
  <NuGetPackageId>MyPackage</NuGetPackageId>
  <NuGetPackageVersion>1.0.0</NuGetPackageVersion>
  <NuGetTargetFramework>.Net Framework,Version=4.6</NuGetTargetFramework>
</Reference>
<Reference Include="netstandard20support\Beta.dll" Condition="'$(TargetFramework)' == 'netstandard2.0'">
  <NuGetPackageId>MyPackage</NuGetPackageId>
  <NuGetPackageVersion>1.0.0</NuGetPackageVersion>
  <NuGetTargetFramework>.Net Standard,Version=2.0</NuGetTargetFramework>
</Reference>
```

Is the project system even going to see both references, or only one at a time depending on the "active" target framework? The end result being that in multi-targeting scenarios you might only see one or the other assembly in Solution Explorer, not both.

We don't have this problem with `FileDefinition` and `FileDependency` items.

nguerrera replied:
The references from NetStandard.Library are already conditional:

https://github.com/dotnet/standard/blob/master/netstandard/pkg/targets/netstandard/NETStandard.Library.targets#L6-L8
https://github.com/dotnet/standard/blob/master/netstandard/pkg/targets/NETStandard.Library.targets#L7

This is *why* the package can't use standard nuget conventions. It wants to add references only to .NETStandard builds, but not to builds targeting a framework that is merely compatible with .NETStandard.

srivatsn commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -182,6 +191,82 @@ private void PopulateAssemblies()
+                }
+
+                var versionComponents = packageVersion.Split('.');
+                if (versionComponents[0] != "2" ||
```
Why make it specific to 2.0?

tmeschter replied:
At the moment I'm assuming a future version will do something different and I don't want to conflict with that. If we turn this into the general solution, then this is not needed.

ericstj replied:
That's a bad assumption.  You wouldn't need to assume if you asked the folks who owned the package.  Now we need to come ask for a fix when we make NETStandard.Library 2.1?

tmeschter commented:
@nguerrera @srivatsn I think a proper general solution is really a feature request at this point, and out of scope for Dev15.3. Can we take this workaround as-is for 15.3 and figure out a general solution later?

tmeschter committed:
Minor comment update

`FileDefinition` and `FileDependency` items may not be the long-term
general solution for handling assemblies that can't be read from the
assets file for some reason, so don't indicate that in the comments.

nguerrera commented:
>  I think a proper general solution is really a feature request at this point, and out of scope for Dev15.3. Can we take this workaround as-is for 15.3 and figure out a general solution later?

I'm OK with that.

tmeschter commented:
@nguerrera 

>>  I think a proper general solution is really a feature request at this point, and out of scope for Dev15.3. Can we take this workaround as-is for 15.3 and figure out a general solution later?

>I'm OK with that.

Could you update your review?

tmeschter commented:
Tagging @livarcocc and @MattGertz for approval.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -182,6 +191,81 @@ private void PopulateAssemblies()
+
+                // This is meant only as a workaround for an issue in NETStandard.Library 2.0.*.
+                // Other packages should use the more general mechanism of creating
+                // FileDefinition and FileDependency items rather than Reference items.
```
There's still a suggestion of FileDefinition/FileDependency mutation as a general mechanism here.

tmeschter replied:
Oops. Corrected.

tmeschter committed:
Update another comment

Remove another reference to `FileDependency` and `FileDefinition`.

natidea commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -182,6 +191,79 @@ private void PopulateAssemblies()
+                }
+
+                // This is meant only as a workaround for an issue in NETStandard.Library 2.0.*.
+                if (packageName != "NETStandard.Library")
```
>"NETStandard.Library" [](start = 35, length = 21)

String comparison ordinal ignore case for robustness

tmeschter replied:
Fixed.

natidea commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -440,6 +522,26 @@ private class AssemblyMetadata : PackageMetadata
+                      version: string.Empty,
+                      path: path,
+                      resolved: true,
+                      isImplicitlyDefined: false)
```
>isImplicitlyDefined: false [](start = 22, length = 26)

Should we regard these assemblies as implicitly defined?

tmeschter replied:
@natidea Possibly. What does "implicitly defined" mean? :-)

natidea replied:
Good question! I think the intent is that references that are not directly in the project file are implicit. Looking through the dependency tree code, I think the relevant code block is [DependencyModel.cs:L41-44](https://github.com/dotnet/project-system/blob/b82639ebeff7912b950ac6e85deaf99069eb27dd/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Models/DependencyModel.cs#L41-L44): implicit nodes do not have a "Remove" action in the context menu.

---
In reply to: [122074507](https://github.com/dotnet/sdk/pull/1330#discussion_r122074507) [](ancestors = 122074507)

nguerrera replied:
@abpiskunov 

IIRC, this controls whether the item can be deleted via IDE. However, there was some discussion which I think was never resolved around the fact that essentially anything in a target cannot be deleted in IDE and relying on metadata for it is problematic.

srivatsn replied:
There is code that was added in VS to see if an item is defined in the project and if not then make it unremovable. So not sure if this is still used.

tmeschter committed:
Fix string comparison

Package names should be compared in a case-insensitive manner.

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -182,6 +191,79 @@ private void PopulateAssemblies()
+                }
+
+                // This is meant only as a workaround for an issue in NETStandard.Library 2.0.*.
+                if (!packageName.Equals("NETStandard.Library", StringComparison.OrdinalIgnoreCase))
```
I don't agree with special casing here.  You already have the PackageIds that you are considering SDK references you should use those.

ericstj replied:
Moreover you can further short-circuit this loop by examining some metadata that is "normally" available.  I'd much more prefer this over the special casing.

```
if (DependencyType.Unknown == GetDependencyType(reference.GetMetadata(MetadataKeys.Type)))
{
    continue;
}
```

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/PreprocessPackageDependenciesDesignTime.cs:
```diff
@@ -182,6 +191,79 @@ private void PopulateAssemblies()
+                    continue;
+                }
+
+                // This is meant only as a workaround for an issue in NETStandard.Library 2.0.*.
```
This is not meant as a workaround.  What is it working around?  This needs to be thought of as a permanent feature.
